debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/FindAllReadsCommand.java

```diff
@@ -0,0 +1,115 @@
+ *******************************************************************************/
+package com.ibm.j9ddr.vm29.tools.ddrinteractive.commands;
+
+import java.io.PrintStream;
+
+import com.ibm.j9ddr.CorruptDataException;
+import com.ibm.j9ddr.tools.ddrinteractive.Command;
+import com.ibm.j9ddr.tools.ddrinteractive.CommandUtils;
+import com.ibm.j9ddr.tools.ddrinteractive.Context;
+import com.ibm.j9ddr.tools.ddrinteractive.DDRInteractiveCommandException;
+import com.ibm.j9ddr.util.PatternString;
+import com.ibm.j9ddr.vm29.j9.DataType;
+import com.ibm.j9ddr.vm29.j9.HashTable;
+import com.ibm.j9ddr.vm29.j9.ModuleHashTable;
+import com.ibm.j9ddr.vm29.j9.ObjectModel;
+import com.ibm.j9ddr.vm29.j9.Pool;
+import com.ibm.j9ddr.vm29.j9.SlotIterator;
+import com.ibm.j9ddr.vm29.pointer.generated.J9HashTablePointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9JavaVMPointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9ModulePointer;
+import com.ibm.j9ddr.vm29.pointer.generated.J9PoolPointer;
+import com.ibm.j9ddr.vm29.pointer.helper.J9ObjectHelper;
+import com.ibm.j9ddr.vm29.pointer.helper.J9RASHelper;
+import com.ibm.j9ddr.vm29.tools.ddrinteractive.JavaVersionHelper;
+
+/**
+ * FindAllReads command displays all modules that read the target module
+ * 
+ * Example:
+ * 			!findallreads 0x00000130550DBB58
+ * Example output: 
+ * 			java.xml			!j9module 0x000001305F46BCB8
+ * 			java.base			!j9module 0x00000130550C7E88
+ *			Found 2 module(s) that read(s) from !j9module	0x00000130550DBB58
+ */
+public class FindAllReadsCommand extends Command{
+
+	public FindAllReadsCommand()
+	{
+		addCommand("findallreads", "<targetModuleAddress>", "find all modules that read target module");
+	}
+
+	public void run(String command, String[] args, Context context, PrintStream out) throws DDRInteractiveCommandException {
+		if (args.length != 1) {
+			CommandUtils.dbgPrint(out, "Usage: !findallreads <targetModuleAddress>\n");
+			return;
+		}
+		try {
+			J9JavaVMPointer vm = J9RASHelper.getVM(DataType.getJ9RASPointer());
+			if (JavaVersionHelper.ensureJava9AndUp(vm, out)) {
+				J9PoolPointer pool = vm.modularityPool();
+				Pool<J9ModulePointer> modulePool = Pool.fromJ9Pool(pool, J9ModulePointer.class);
+				SlotIterator<J9ModulePointer> poolIterator = modulePool.iterator();
+				J9ModulePointer modulePtr = null;
+
+				int hitCount = 0;
+				String searchModuleAddress = args[0];
+				PatternString pattern = new PatternString(searchModuleAddress);
+
+				while (poolIterator.hasNext()) {
+					modulePtr = poolIterator.next();
+					/*
+					 * Since both packages and modules could be in the pool, the
+					 * iterator checks on the region and make sure that it is
+					 * pointing to a module. (For modules, the first field must
+					 * be in heap since it is an object while for j9package it
+					 * is a J9UTF8)
+					 */
+					if (ObjectModel.isPointerInHeap(vm, modulePtr.moduleName())) {
```

Comment:
I only used the module name for the `printf` below so it prints the name of the module because the output format is like that for other commands. I can remove if it seems extra

Reply:
Why is condition `ObjectModel.isPointerInHeap(vm, modulePtr.moduleName())` instead of just `ObjectModel.isPointerInHeap(vm, modulePtr)`?

Reply:
The `modulePool` contains J9Modules and J9Packages, neither of which are on the java heap.

Reply:
I see. Is this the only way we can distinguish modules from packages? It seems somewhat fragile to me.

Reply:
The alternative is to walk all classloaders, and then walk all modules in those classloaders.

Reply:
I think that's a better approach: we don't want this to break someday just because package names moved to the heap.

Reply:
I changed all the commands related to this to using class loaders.